### PR TITLE
Add build methods to create CompletionSuggestionBuilder from query and parse contexts

### PR DESCRIPTION
**Pull request by areek**
This implements methods in CompletionSuggestionBuilder to construct a builder given query and parse contexts respectively. Now we ensure that the builder outputs identical output when used directly or through json parsing by storing query contexts as binary in the builder. All validation requiring `MapperService` are done once the suggestion is built on the shard level.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java [hunk 1]
```diff
@@ -74,27 +67,20 @@
...
-        REGEXP_PARSER.declareInt(RegexOptions.Builder::setMaxDeterminizedStates, RegexOptions.MAX_DETERMINIZED_STATES);
-        REGEXP_PARSER.declareStringOrNull(RegexOptions.Builder::setFlags, RegexOptions.FLAGS_VALUE);
-
         TLP_PARSER.declareStringArray(CompletionSuggestionContext::setPayloadFields, CompletionSuggestionBuilder.PAYLOAD_FIELD);
-        TLP_PARSER.declareObjectOrDefault(CompletionSuggestionContext::setFuzzyOptionsBuilder, FUZZY_PARSER, FuzzyOptions.Builder::new, FuzzyOptions.FUZZY_OPTIONS);
-        TLP_PARSER.declareObject(CompletionSuggestionContext::setRegexOptionsBuilder, REGEXP_PARSER, RegexOptions.REGEX_OPTIONS);
+        TLP_PARSER.declareField((parser, completionSuggestionContext, context) -> {
+                if (parser.currentToken() == XContentParser.Token.VALUE_BOOLEAN) {
+                    if (parser.booleanValue()) {
+                        completionSuggestionContext.setFuzzyOptions(new FuzzyOptions.Builder().build());
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java]: Out of curiosity, I tried to find where this was covered before, but couldn't, maybe you can point me to it?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 2]
```diff
@@ -117,35 +167,38 @@ public CompletionSuggestionBuilder regex(String regex, RegexOptions regexOptions
...
     }
 
-    private CompletionSuggestionBuilder contexts(String name, QueryContext... queryContexts) {
-        List<QueryContext> contexts = this.queryContexts.get(name);
-        if (contexts == null) {
-            contexts = new ArrayList<>(2);
-            this.queryContexts.put(name, contexts);
-        }
-        Collections.addAll(contexts, queryContexts);
+    private String field;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: This looks like its only used in innerFromXContent, to circumvent the limitations imposed by ObjectParser and requiring the fieldname at construction time. Can we encapsulate this (and the setter) in an `InnerBuilder extend CompletionSuggestionBuilder` and then use that as the value object the parser fills? Otherwise this extra field feels a bit odd, given that it is not serialized etc... (even though it is private).

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 3]
```diff
@@ -117,35 +167,38 @@ public CompletionSuggestionBuilder regex(String regex, RegexOptions regexOptions
      * Note: Only doc values enabled fields are supported
      */
     public CompletionSuggestionBuilder payload(List<String> fields) {
-        this.payloadFields.addAll(fields);
+        this.payloadFields = fields;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: nit: now that payloadFields is replaces, can you add a null-check?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/RegexOptions.java [hunk 4]
```diff
@@ -35,10 +38,26 @@
...
+    private static final ParseField MAX_DETERMINIZED_STATES = new ParseField("max_determinized_states");
+
+    private static ObjectParser<Builder, Void> PARSER = new ObjectParser<>(REGEX_OPTIONS.getPreferredName(), Builder::new);
+    static {
+        PARSER.declareInt(Builder::setMaxDeterminizedStates, MAX_DETERMINIZED_STATES);
+        PARSER.declareField((parser, builder, aVoid) -> {
+            if (parser.currentToken() == XContentParser.Token.VALUE_STRING) {
+                builder.setFlags(parser.text());
+            } else if (parser.currentToken() == XContentParser.Token.VALUE_NUMBER) {
+                builder.setFlagsValue(parser.intValue());
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/RegexOptions.java]: Ah nice, this also wasn't covered before when this was part of CompletionSuggestionParser?

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java [hunk 5]
```diff
@@ -211,6 +213,22 @@ public void testFromXContent() throws IOException {
         }
     }
 
+    protected Tuple<MapperService, SB> mockMapperServiceAndSuggestionBuilder(
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I don't understand the use of this helper method, it only seems to be used internally and by introducing it I think the code gets harder to read. Might just be me though.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 6]
```diff
@@ -117,35 +167,38 @@ public CompletionSuggestionBuilder regex(String regex, RegexOptions regexOptions
...
     /**
-     * Sets query contexts for a category context
-     * @param name of the category context to execute on
-     * @param queryContexts a list of {@link CategoryQueryContext}
+     * Sets query contexts for completion
+     * @param queryContexts named query contexts
+     *                      see {@link org.elasticsearch.search.suggest.completion.context.CategoryQueryContext}
+     *                      and {@link org.elasticsearch.search.suggest.completion.context.GeoQueryContext}
      */
-    public CompletionSuggestionBuilder categoryContexts(String name, CategoryQueryContext... queryContexts) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Was this setter just introduced with the new completion suggester on master or is this a change in the java api that we should track in the migrate.asciidoc?

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java (hunk 5 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: Nevermind, I just got to the part where this is overwritten in CompletionSuggesterBuilderTests. Quiet tricky though, I wonder if this could be made somewhat easier to understand, haven't completely wrapped my head around it yet.

**Comment by cbuescher** [comment]
@areek I did a round of review and left a few questions and comments. I assume there is no alternative to going back to parsing the query contexts on the shard. I wonder if there's an easier way to mock the mapper service in the tests, I found that part a bit hard to follow. Other than that I think it looks great.

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java (hunk 5 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I took a quick stab at sth. that avoids the Tuple and replacing the test builder [here](https://gist.github.com/cbuescher/88531fe7c2abd38936ef), but it still looks a little bit strange to me.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java (hunk 1 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java]: This was covered in https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java#L99.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 2 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Thanks for the suggestion! encapsulated the field setter in `InnerBuilder`

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 3 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: Thanks! added

File: core/src/main/java/org/elasticsearch/search/suggest/completion/RegexOptions.java (hunk 4 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/RegexOptions.java]: no this was not covered before :)

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java (hunk 6 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: These setters were introduced with the new completion suggester on master

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java]: thanks

File: core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java (hunk 5 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I had another go at this (thought I posted the link already yesterday, but aparently forgot). Choosing the random geo/context names upfront makes the BuilderAndInfo go away and the whole thin slightly easier I think, but I'm fine either way. If you are interested, I tried something like [this](https://gist.github.com/cbuescher/34cf78ec6f10bab16fbb).

Comment (areek) [inline core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java]: I took another stab at simplifying [this](https://github.com/areek/elasticsearch/commit/3940c898422eb4d2f031cf05dc16f1bdd6d7978f), Thoughts?

**Comment by areek** [comment]
Thanks @cbuescher for the feedback, I addressed all the comments!

**Comment by cbuescher** [comment]
Looks great, by which I mean LGTM.
